File: hw2/prob10_2.py
# GLOBAL ALIGNMENT IN LINEAR SPACE PROBLEM

# Find the highest-scoring alignment between two strings using a scoring matrix in linear space.

# Given:  Two long amino acid strings (of length approximately 10,000).
# Return: The maximum alignment score of these strings, followed by an alignment achieving this maximum score. Use the BLOSUM62 scoring matrix and indel penalty = 5.

import sys

class BLOSUM62( object ):
    def __init__(self):
        with open('BLOSUM62.txt') as input:
            items = [line.strip().split() for line in input.readlines()]
            self.m = {(item[0], item[1]):int(item[2]) for item in items}

    def __getitem__(self, pair):
        return self.m[pair[0], pair[1]]

def linearSpaceGlobalAlignment( v, w, matrix, sigma ):

    def middleColumnScore( v, w, matrix, sigma ):
        S = [[i*j*sigma for j in xrange(-1, 1)] for i in xrange(len(v)+1)]
        S[0][1] = -sigma
        backtrack = [0]*(len(v)+1)
        for j in xrange(1, len(w)/2+1):
            for i in xrange(0, len(v)+1):
                if i == 0:
                    S[i][1] = -j*sigma
                else:
                    scores = [S[i-1][0] + matrix[v[i-1], w[j-1]], S[i][0] - sigma, S[i-1][1] - sigma]
                    S[i][1] = max(scores)
                    backtrack[i] = scores.index(S[i][1])
            if j != len(w)/2:
                S = [[row[1]]*2 for row in S]
        return [row[1] for row in S], backtrack

    def middleEdge( v, w, matrix, sigma ):
        sourceToMiddle = middleColumnScore(v, w, matrix, sigma)[0]
        middleToSink, backtrack = map(lambda l: l[::-1], middleColumnScore(v[::-1], w[::-1]+['', '$'][len(w) % 2 == 1 and len(w) > 1], matrix, sigma))
        scores = map(sum, zip(sourceToMiddle, middleToSink))
        maxMiddle = max(xrange(len(scores)), key=lambda i: scores[i])
        if maxMiddle == len(scores) - 1:
            nextNode = (maxMiddle, len(w)/2 + 1)
        else:
            nextNode = [(maxMiddle + 1, len(w)/2 + 1), (maxMiddle, len(w)/2 + 1), (maxMiddle + 1, len(w)/2),][backtrack[maxMiddle]]
        return (maxMiddle, len(w)/2), nextNode

    def globalAlignment( v, w, matrix, sigma ):
        S = [[0 for repeat_j in xrange(len(w)+1)] for repeat_i in xrange(len(v)+1)]
        backtrack = [[0 for repeat_j in xrange(len(w)+1)] for repeat_i in xrange(len(v)+1)]
        for i in xrange(1, len(v)+1):
            S[i][0] = -i*sigma
        for j in xrange(1, len(w)+1):
            S[0][j] = -j*sigma
        for i in xrange(1, len(v)+1):
            for j in xrange(1, len(w)+1):
                scores = [S[i-1][j] - sigma, S[i][j-1] - sigma, S[i-1][j-1] + matrix[v[i-1], w[j-1]]]
                S[i][j] = max(scores)
                backtrack[i][j] = scores.index(S[i][j])
        insert_indel = lambda word, i: word[:i] + '-' + word[i:]
        vAligned, wAligned = v, w
        i, j = len(v), len(w)
        maxScore = str(S[i][j])
        while i*j != 0:
            if backtrack[i][j] == 0:
                i -= 1
                wAligned = insert_indel(wAligned, j)
            elif backtrack[i][j] == 1:
                j -= 1
                vAligned = insert_indel(vAligned, i)
            else:
                i -= 1
                j -= 1
        for repeat in xrange(i):
            wAligned = insert_indel(wAligned, 0)
        for repeat in xrange(j):
            vAligned = insert_indel(vAligned, 0)
        return maxScore, vAligned, wAligned

    def linearSpaceAlignment( top, bottom, left, right ):
        if left == right:
            return [v[top:bottom], '-'*(bottom - top)]
        elif top == bottom:
            return ['-'*(right - left), w[left:right]]
        elif bottom - top == 1 or right - left == 1:
            return globalAlignment(v[top:bottom], w[left:right], matrix, sigma)[1:]
        else:
            midNode, nextNode = middleEdge(v[top:bottom], w[left:right], matrix, sigma)
            midNode = tuple(map(sum, zip(midNode, [top, left])))
            nextNode = tuple(map(sum, zip(nextNode, [top, left])))
            current = [['-', v[midNode[0] % len(v)]][nextNode[0] - midNode[0]], ['-', w[midNode[1] % len(w)]][nextNode[1] - midNode[1]]]
            A = linearSpaceAlignment(top, midNode[0], left, midNode[1])
            B = linearSpaceAlignment(nextNode[0], bottom, nextNode[1], right)
            return [A[i] + current[i] + B[i] for i in xrange(2)]
    vAligned, wAligned = linearSpaceAlignment(0, len(v), 0, len(w))
    score = sum([-sigma if '-' in pair else matrix[pair] for pair in zip(vAligned, wAligned)])
    return str(score), vAligned, wAligned

if __name__ == '__main__':
    with open(sys.argv[1]) as input:
        word1, word2 = [line.strip() for line in input.readlines()]
    alignment = linearSpaceGlobalAlignment(word1, word2, BLOSUM62(), 5)
    print '\n'.join(alignment)

File: hw2/prob10_2.java
/**
 *  GLOBAL ALIGNMENT PROBLEM
 *
 *  Find the highest-scoring alignment between two strings using a scoring matrix.
 *
 *  Given:  Two protein strings.
 *  Return: The maximum alignment score of these strings followed by an alignment achieving this maximum score. Use the BLOSUM62 scoring matrix and indel penalty sigma = 5. (If multiple alignments achieving the maximum score exist, you may return any one.)
 */

import java.util.*;
public class GlobalAlignmentProblem {
    public static String aminosAlpha = "ACDEFGHIKLMNPQRSTVWY";
    static int[][] blosum62 = {{4,0,-2,-1,-2,0,-2,-1,-1,-1,-1,-2,-1,-1,-1,1,0,0,-3,-2},{0,9,-3,-4,-2,-3,-3,-1,-3,-1,-1,-3,-3,-3,-3,-1,-1,-1,-2,-2},{-2,-3,6,2,-3,-1,-1,-3,-1,-4,-3,1,-1,0,-2,0,-1,-3,-4,-3},{-1,-4,2,5,-3,-2,0,-3,1,-3,-2,0,-1,2,0,0,-1,-2,-3,-2},{-2,-2,-3,-3,6,-3,-1,0,-3,0,0,-3,-4,-3,-3,-2,-2,-1,1,3},{0,-3,-1,-2,-3,6,-2,-4,-2,-4,-3,0,-2,-2,-2,0,-2,-3,-2,-3},{-2,-3,-1,0,-1,-2,8,-3,-1,-3,-2,1,-2,0,0,-1,-2,-3,-2,2},{-1,-1,-3,-3,0,-4,-3,4,-3,2,1,-3,-3,-3,-3,-2,-1,3,-3,-1},{-1,-3,-1,1,-3,-2,-1,-3,5,-2,-1,0,-1,1,2,0,-1,-2,-3,-2},{-1,-1,-4,-3,0,-4,-3,2,-2,4,2,-3,-3,-2,-2,-2,-1,1,-2,-1},{-1,-1,-3,-2,0,-3,-2,1,-1,2,5,-2,-2,0,-1,-1,-1,1,-1,-1},{-2,-3,1,0,-3,0,1,-3,0,-3,-2,6,-2,0,0,1,0,-3,-4,-2},{-1,-3,-1,-1,-4,-2,-2,-3,-1,-3,-2,-2,7,-1,-2,-1,-1,-2,-4,-3},{-1,-3,0,2,-3,-2,0,-3,1,-2,0,0,-1,5,1,0,-1,-2,-2,-1},{-1,-3,-2,0,-3,-2,0,-3,2,-2,-1,0,-2,1,5,-1,-1,-3,-3,-2},{1,-1,0,0,-2,0,-1,-2,0,-2,-1,1,-1,0,-1,4,1,-2,-3,-2},{0,-1,-1,-1,-2,-2,-2,-1,-1,-1,-1,0,-1,-1,-1,1,5,0,-2,-2},{0,-1,-3,-2,-1,-3,-3,3,-2,1,1,-3,-2,-2,-3,-2,0,4,-3,-1},{-3,-2,-4,-3,1,-2,-2,-3,-3,-2,-1,-4,-4,-2,-3,-3,-2,-3,11,2},{-2,-2,-3,-2,3,-3,2,-1,-2,-1,-1,-2,-3,-1,-2,-2,-2,-1,2,7}};

    public static void main( String[] args ) {
        // DEFINE "s1" AND "s2"!!!
        String s1 = "";
        String s2 = "";
        // DEFINE "s1" AND "s2"!!!
        int indel = 5;
        String[] out = proteinGlobalAlignment(s1,s2,indel);
        System.out.println(out[0]);
        System.out.println(out[1]);
        System.out.println(out[2]);
    }

    public static String[] proteinGlobalAlignment( String s, String t, int indel ) {
        // Uses BLOSUM62. PASS indel AS POSITIVE INTEGER, NOT NEGATIVE!
        int[][] S = new int[s.length()+1][t.length()+1];
        int[][] opt = new int[s.length()+1][t.length()+1]; // 1 = right, 2 = down, 3 = diag
        S[0][0] = 0;
        for(int i = 1; i <= s.length(); ++i) {
            S[i][0] = S[i-1][0] - indel;
            opt[i][0] = 1;
        }
        for(int j = 1; j <= t.length(); ++j) {
            S[0][j] = S[0][j-1] - indel;
            opt[0][j] = 2;
        }
        for(int i = 1; i <= s.length(); ++i) {
            for(int j = 1; j <= t.length(); ++j) {
                int opt1 = S[i][j-1] - indel; // right
                int opt2 = S[i-1][j] - indel; // down
                int opt3 = S[i-1][j-1] + blosum62[aminosAlpha.indexOf(s.charAt(i-1))][aminosAlpha.indexOf(t.charAt(j-1))]; // diag
                S[i][j] = opt1;
                opt[i][j] = 1;
                if(opt2 > S[i][j]) {
                    S[i][j] = opt2;
                    opt[i][j] = 2;
                }
                if(opt3 > S[i][j]) {
                    S[i][j] = opt3;
                    opt[i][j] = 3;
                }
            }
        }
        String[] out = new String[3];
        out[0] = "" + S[s.length()][t.length()];
        out[1] = "";
        out[2] = "";
        int i = s.length();
        int j = t.length();
        while(i > 0 && j > 0) {
            if(opt[i][j] == 1) { // right
                out[1] = '-' + out[1];
                out[2] = t.charAt(j-- - 1) + out[2];
            }
            else if(opt[i][j] == 2) { // down
                out[1] = s.charAt(i-- - 1) + out[1];
                out[2] = '-' + out[2];
            }
            else if(opt[i][j] == 3) { // diag
                out[1] = s.charAt(i-- - 1) + out[1];
                out[2] = t.charAt(j-- - 1) + out[2];
            }
        }
        if(i > 0) {
            out[1] = s.substring(0,i) + out[1];
            String add = "";
            for(int x = 0; x < i; ++x) {
                add += '-';
            }
            out[2] = add + out[2];
        }
        if(j > 0) {
            out[2] = t.substring(0,j) + out[2];
            String add = "";
            for(int x = 0; x < j; ++x) {
                add += '-';
            }
            out[1] = add + out[1];
        }
        return out;
    }
}
